editor/src/core/shared/project-file-types.ts

```diff
@@ -190,30 +240,60 @@ export const enum RevisionsState {
   BothMatch = 'BOTH_MATCH',
 }
 
-export interface UIJSFile {
-  type: 'UI_JS_FILE'
-  fileContents: ParseResult
-  lastSavedContents: ParseResult | null // it is null when the file is saved
+export interface TextFileContents {
+  code: string
+  parsed: ParsedTextFile
   revisionsState: RevisionsState
+}
+
+export function textFileContents(
+  code: string,
+  parsed: ParsedTextFile,
+  revisionsState: RevisionsState,
+): TextFileContents {
+  return {
+    code: code,
+    parsed: parsed,
+    revisionsState: revisionsState,
+  }
+}
+
+export interface TextFile {
+  type: 'TEXT_FILE'
+  fileContents: TextFileContents
+  lastSavedContents: TextFileContents | null // it is null when the file is saved
   lastRevisedTime: number
 }
 
-export function isUIJSFile(projectFile: ProjectFile | null): projectFile is UIJSFile {
-  return projectFile != null && projectFile.type === 'UI_JS_FILE'
+export function textFile(
+  fileContents: TextFileContents,
+  lastSavedContents: TextFileContents | null,
+  lastRevisedTime: number,
+): TextFile {
+  return {
+    type: 'TEXT_FILE',
+    fileContents: fileContents,
+    lastSavedContents: lastSavedContents,
+    lastRevisedTime: lastRevisedTime,
+  }
 }
 
-export interface CodeFile {
-  type: 'CODE_FILE'
-  fileContents: string // Maybe should be more like an asset?
-  lastSavedContents: string | null // it is null when the file is saved
+export function codeFile(fileContents: string, lastSavedContents: string | null): TextFile {
+  return textFile(
+    textFileContents(fileContents, unparsed, RevisionsState.CodeAhead),
+    lastSavedContents == null
+      ? null
+      : textFileContents(lastSavedContents, unparsed, RevisionsState.CodeAhead),
+    0,
+  )
 }
 
-export function isCodeFile(projectFile: ProjectFile | null): projectFile is CodeFile {
-  return projectFile != null && projectFile.type === 'CODE_FILE'
+export function isTextFile(projectFile: ProjectFile | null): projectFile is TextFile {
+  return projectFile != null && projectFile.type === 'TEXT_FILE'
 }
 
-export function isCodeOrUiJsFile(file: ProjectFile): file is CodeFile | UIJSFile {
-  return file.type === 'CODE_FILE' || file.type === 'UI_JS_FILE'
+export function isParsedTextFile(projectFile: ProjectFile | null): projectFile is TextFile {
+  return isTextFile(projectFile) && !isUnparsed(projectFile.fileContents.parsed)
```

Comment:
what if I have a parseable text file that is not parsed yet? can we use this to distinguish files that are parseable from files we don't intend to parse? is that even an important distinction?

Reply:
ping Rheese

Reply:
This was something we thought about but realised that in doing so we'd end up in a situation where we would be marking files as "never parse", which would end up with us also having to build in logic to handle those files changing in a way that would allow us to parse them.

So for example, say a user creates a file `thing.json`, we would want to mark that as "never parse". But they named it that by accident, because really they wanted to call it `thing.js`, so now we have to make sure that renaming to a supported file extension triggers the parsing. It's not a huge issue, but means we have to build in logic to cover edge cases that would otherwise be handled for free.